### PR TITLE
Custom column encoding for benchmarks

### LINKED ISSUE #861: Add column encoding option for benchmarks
Operator implementations are often created with compressed chunks in mind. However, `hyriseBenchmarkTPCH` for example creates test tables without chunk encoding. Since these tables are generated at program startup, it would be nice to have an option like **`--encoding`** taking an `EncodingType` enum as argument, defaulting to `Unencoded` or whatever. Thoughts?

### PR DESCRIPTION
**Pull request by lawben**
This PR adds the option to encode specific columns in benchmarks.

Taken from code:

Represents the column encodings specified for a benchmark. If encoding (and vector compression) were specified via command line args, this will contain no custom encoding mapping but only the column default. This will lead to each column in each chunk to be encoded/compressed by this default. If a JSON config was provided, a column specific encoding/compression can be chosen (same in each chunk). The JSON config must look like this:

All encoding/compression types can be viewed with the `help` command or seen in constant_mappings.cpp. The encoding is always required, the compression is optional.
```
{
  "default": {
    "encoding": <ENCODING_TYPE_STRING>,               // required
    "compression": <VECTOR_COMPRESSION_TYPE_STRING>,  // optional
  },
  "custom": {
    <TABLE_NAME>: {
      <COLUMN_NAME>: {
        "encoding": <ENCODING_TYPE_STRING>,
        "compression": <VECTOR_COMPRESSION_TYPE_STRING>
      },
      <COLUMN_NAME>: {
        "encoding": <ENCODING_TYPE_STRING>
      }
    },
    <TABLE_NAME>: {
      <COLUMN_NAME>: {
        "encoding": <ENCODING_TYPE_STRING>,
        "compression": <VECTOR_COMPRESSION_TYPE_STRING>
      }
    }
  }
}
```

Closes #861.

### CODE REVIEW

**Comment by lawben** [comment]
@Bouncner @adi64 comments? as you were the ones to suggest this

**Comment by Bouncner** [comment]
Cannot test right now, but look superb!

**Comment by mrks** [comment]
Neat. I'll do a code review in a bit, but one question:
How much work would it be to support per-column-type-defaults? Right now, we could not use FrameOfReference as a default because it only supports ints.

**Comment by lawben** [comment]
@mrks Absolutely doable :) I would just need a hint on how the config for that would look like? 
```
{
  "default": { "encoding": ... },
  "int_default": { "encoding": "FrameOfReference", ... },
  "string_default: { "encoding": "Dictionary, ... }
  "custom": ...
}
```

Something like this?

**Comment by mrks** [comment]
My first idea would be something like this, but I could live with your version as well.

```
root
+- default
  +- encoding
  +- (compression)
+- int
  +- encoding: FrameOfReference
+- TableName.ColumnName
  +- encoding:
...
```

Open for discussion.

**Comment by lawben** [comment]
I prefer yours. But lets move that to a separate PR just to keep changes small.

**Comment by adi64** [comment]
Will look into it tomorrow :)

File: src/benchmarklib/benchmark_utils.hpp [hunk 1]
```diff
@@ -24,6 +25,8 @@ using TimePoint = std::chrono::high_resolution_clock::time_point;
 using NamedQuery = std::pair<std::string, std::string>;
 using NamedQueries = std::vector<NamedQuery>;
 
+using EncodingMapping = std::unordered_map<std::string, std::map<std::string, ColumnEncodingSpec>>;
```
Comment (adi64) [inline src/benchmarklib/benchmark_utils.hpp]: This maps table name -> column name -> encoding spec, right? I'd love to have a name here that tells me that.
`EncodingSpecByTableAndColumn` would get that across, but be a horrid name...

File: src/benchmarklib/benchmark_utils.cpp [hunk 2]
```diff
@@ -139,22 +140,23 @@ BenchmarkConfig CLIConfigParser::parse_basic_options_json_config(const nlohmann:
...
+  std::unique_ptr<EncodingConfig> encoding_config{};
+  const auto encoding_type_str = json_config.value("encoding", "Dictionary");
+  const auto compression_type_str = json_config.value("compression", "");
+  if (boost::algorithm::ends_with(encoding_type_str, ".json")) {
+    // Use encoding file instead of default type
+    encoding_config = std::make_unique<EncodingConfig>(parse_encoding_config(encoding_type_str));
+    out << "- Encoding is custom from " << encoding_type_str << "" << std::endl;
+
+    if (!compression_type_str.empty()) {
+      out << "- Specified compression type and an encoding file. The compression will be ignored." << std::endl;
```
Comment (mrks) [inline src/benchmarklib/benchmark_utils.cpp]: Specified *both* ...
Also - better fail here?

File: src/benchmarklib/benchmark_utils.hpp (hunk 1 above)
Comment (lawben) [reply src/benchmarklib/benchmark_utils.hpp]: i'll try to come up with something better :)

**Pull request merged**